src/rgw/rgw_data_sync.cc

```diff
@@ -1413,6 +1413,214 @@ class RGWDataSyncSingleEntryCR : public RGWCoroutine {
   }
 };
 
+class RGWFullSyncErrorRepoCR: public RGWCoroutine {
+  RGWDataSyncCtx *sc;
+  RGWDataSyncEnv *sync_env;
+  rgw_bucket_shard source_bs;
+  std::string error_marker;
+  ceph::real_time timestamp;
+  RGWSyncTraceNodeRef tn;
+  rgw_bucket_index_marker_info remote_info;
+  rgw_pool pool;
+  RGWDataChangesLog *datalog_changes{nullptr};
+  rgw_raw_obj error_repo;
+  uint32_t sid;
+  rgw_bucket_shard bs;
+  std::vector<store_gen_shards>::const_iterator each;
+
+  rgw_raw_obj datalog_oid_for_error_repo(rgw_bucket_shard& bs) {
+    int datalog_shard = datalog_changes->choose_oid(bs);
+    string oid = RGWDataSyncStatusManager::shard_obj_name(sc->source_zone, datalog_shard);
+    return rgw_raw_obj(pool, oid + ".retry");
+  }
+
+public:
+  RGWFullSyncErrorRepoCR(RGWDataSyncCtx *_sc, rgw_bucket_shard& _source_bs, std::string _error_marker,
+                         ceph::real_time& _timestamp, RGWSyncTraceNodeRef& _tn)
+    : RGWCoroutine(_sc->cct), sc(_sc), sync_env(_sc->env),
+      source_bs(_source_bs), error_marker(_error_marker), timestamp(_timestamp), tn(_tn) {
+        tn = sync_env->sync_tracer->add_node(_tn, "error_repo", SSTR(bucket_shard_str(source_bs)));
+      }
+
+  int operate(const DoutPrefixProvider *dpp) override {
+    reenter(this) {
+      yield call(new RGWReadRemoteBucketIndexLogInfoCR(sc, source_bs.bucket, &remote_info));
+      if (retcode < 0) {
+        return set_cr_error(retcode);
+      }
+
+      each = remote_info.generations.cbegin();
+      for (; each != remote_info.generations.cend(); each++) {
+        for (sid = 0; sid < each->num_shards; sid++) {
+          bs.bucket = source_bs.bucket;
+          bs.shard_id = sid;
+          error_repo = datalog_oid_for_error_repo(bs);
+          tn->log(10, SSTR("writing shard_id " << sid << " of gen " << each->gen << " to error repo for retry"));
+          yield_spawn_window(rgw::error_repo::write_cr(sync_env->store->svc()->rados, error_repo,
+                            rgw::error_repo::encode_key(bs, each->gen),
+                            timestamp), cct->_conf->rgw_data_sync_spawn_window,
+                            [&](uint64_t stack_id, int ret) {
+                              if (ret < 0) {
+                                tn->log(10, SSTR("writing to error repo returned error: " << ret));
+                              }
+                              return ret;
+                            });
+        }
+      }
+      drain_all_cb([&](uint64_t stack_id, int ret) {
+                   if (ret < 0) {
+                     tn->log(10, SSTR("writing to error repo returned error: " << ret));
+                   }
+                   return ret;
+                 });
+
+      spawn(rgw::error_repo::remove_cr(sync_env->store->svc()->rados, error_repo,
+                                      error_marker, timestamp), false);
+
+      return set_cr_done();
```

Comment:
should we add a retry loop here for write_cr() if _retcode < 0_?

Reply:
`RGWFullSyncErrorRepoCR` is called when we see a 'full sync obligation' in the error repo. if we don't call this `remove_cr()` to remove it, it'll stay in the error repo and we'll retry it later automatically

so we don't need any extra retry logic here, we just need to avoid calling this `remove_cr()` if there were any other failures

Reply:
understood. but don't we need to remove them from the error_repo at some point?

Reply:
yeah sorry, we still need the `remove_cr()` here - it just needs to be conditional on the value of `retcode`